service/frontend/health_check.go

```diff
@@ -0,0 +1,124 @@
+package frontend
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+
+	enumsspb "go.temporal.io/server/api/enums/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/cache"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/membership"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/service/matching"
+)
+
+type (
+	HealthCheckClient struct {
+		clients     []healthpb.HealthClient
+		serviceName string
+		failureRate dynamicconfig.FloatPropertyFn
+	}
+
+	HealthCheckClientFactory struct {
+		membershipMonitor membership.Monitor
+		rpcFactory        common.RPCFactory
+		config            *Config
+		connectionCache   cache.Cache
+	}
+
+	healthCheckResponse struct {
+		*healthpb.HealthCheckResponse
+		err error
+	}
+)
+
+func (c *HealthCheckClient) Check(ctx context.Context) enumsspb.HealthState {
+	respCh := make(chan healthCheckResponse, len(c.clients))
+	for _, client := range c.clients {
+		go func(hc healthpb.HealthClient) {
+			resp, err := hc.Check(ctx, &healthpb.HealthCheckRequest{Service: c.serviceName})
+			respCh <- healthCheckResponse{
+				HealthCheckResponse: resp,
+				err:                 err,
+			}
+		}(client)
+	}
+	close(respCh)
+
+	var failureCount float64
+	for resp := range respCh {
+		if resp.err != nil || resp.GetStatus() != healthpb.HealthCheckResponse_SERVING {
```

Comment:
Should we distinguish "failed to check" from "not serving"?

Reply:
what is the consideration? For those two, it only checks the connection.

Reply:
Ok if it only checks connection.